### ISSUE TITLE
Some questions about Pseudo Text-Feature Generation

### ISSUE THREAD

**Issue by Cwj1212** [drboog/Lafite#11]
Sorry to bother you, I'm having some trouble reading the code

1.Both hyper-parameters ratio and mixing_prob seem to be able to control the generation of pseudo-labels, one is constructed when the dataset is read, and the other is constructed before the forward process. It looks like you are using the hyper-parameter mixing_prob to control the generation of pseudo-labels, I don't know if the a hyper-parameter ratio plays a role in the experiment.It seems that the two methods of constructing pseudo-labels are same.

2.
https://github.com/drboog/Lafite/blob/888efcdd835daa4b54742f4b7398d6a70c69f5e8/training/loss.py#L178
I don't know what the formula corresponding to this comment looks like. It seems to be the lower limit ‘c’ of the similarity between the pseudo-text feature and the image feature given the coefficient of Fixed perturbations. But bringing in 0.75, the value of the formula is negative, is the similarity between the two too low?

3.As mentioned in question 2, I have some confusion about the similarity between features. The first is about CLIP (the generation of pseudo-features depends on the high alignment of text and image in the joint semantic space of CLIP). According to the results of the text encoder and image encoder of CLIP, the cos similarity of image-text pairs is about 0.3, however The similarity between any two text-text, image-image is about 0.5.
This result is very confusing to me. This shows that the pseudo-text feature constructed with image features seem to have very low similarity to their corresponding real-text feature, especially after adding large noise perturbations.Could it be that this noise is doing something special, I can't seem to understand ？The similarity between any two text features is 0.5, but the image feature with a similarity of 0.3 is used to replace it after adding perturbation. It doesn't seem reasonable.
The model is trained under pseudo-text features, but generates corresponding images from real-text features in the inference stage.

**Comment by drboog** [comment]
1. The "ratio" is not used in the experiments for this paper. Basically, I only used mixing_prob=0 which is fully-supervised setting and mixing_prob=1 which is language-free setting.

3. Just ignore this comment, I will delete it. This lower bound is too loose, and I forgot to delete it. If you are interested in the bound for similarity, you can look at the theorem in our paper. Actually you can quickly test that 0.75 will lead to similarity around 0.3, this is why we use 0.75;

5. And I mentioned, the hyper-parameter 0.75 is tuned so that pseudo image-text similarity is around 0.3. If you are interested in image-image similarity, text-text similarity, image-text similarity, I suggest you read this paper [Mind the Gap: Understanding the Modality Gap in Multi-modal Contrastive Representation Learning](https://arxiv.org/abs/2203.02053)

**Comment by Cwj1212** [comment]
Thank you very much for your answer, your answer is always timely and effective. I'm going to read this article.
Before reading the article, I would like to confirm. The similarity of the real text image is about 0.3. After the perturbation. The similarity between the pseudo-text and the image is also about 0.3. So is the similarity between the pseudo-text label and the real text high at this time,? (There is something wrong with my server, so I can only ask you)

**Comment by Cwj1212** [comment]
In the language-free setting, the model learns to generate image from pseudo-text, while the model never seen the real text. The trained model can generate image from real text. How the model understands and extracts information from real text features.

**Comment by drboog** [comment]
> Thank you very much for your answer, your answer is always timely and effective. I'm going to read this article. Before reading the article, I would like to confirm. The similarity of the real text image is about 0.3. After the perturbation. The similarity between the pseudo-text and the image is also about 0.3. So is the similarity between the pseudo-text label and the real text high at this time,? (There is something wrong with my server, so I can only ask you)

"The similarity of the real text image is about 0.3." Yes.
"The similarity between the pseudo-text and the image is also about 0.3" Yes.
"So is the similarity between the pseudo-text and the real text high at this time,?" No.

Because the feature space is 512-dimensional, which is relatively high for perturbation, there is no guarantee that the fake text feature has high similarity with real text feature. But you may assume that we have some probability to generate a fake text feature that is similar to real text feature. During training, we will feed many pseudo text features to the model,  you can regard them as a mixture of features which could be similar and dissimilar to real text feature, or simple  regard them as noised real text features. Because of generalization, the model can learn some information for inference, but of course not as good as learning from real paired text features.

To improve language-free method, what we can think is how to generate "better" fake text features, which is what I'm currently doing.

**Comment by Cwj1212** [comment]
That is to say, the pseudo-text features will be similar to the real text features with a certain probability, and the rest dissimilar can be regarded as noise during the model training process. But I intuitively feel that the probability that the pseudo-text features obtained by the image features plus the perturbation are similar to the real text features seems to be extremely low, because the similarity within the modality is greater than the similarity between the image and text pairs.

**Comment by EmiliaKKK** [comment]
Really thank you for your patient explanation! Now there's a lot of clarity on this! So maybe by using a "Gap-free" CLIP would increase the language-free training efficiency? Since we dont need to add such big noise.
Anyway, hope you success in your future work!

**Comment by drboog** [comment]
> That is to say, the pseudo-text features will be similar to the real text features with a certain probability, and the rest dissimilar can be regarded as noise during the model training process. But I intuitively feel that the probability that the pseudo-text features obtained by the image features plus the perturbation are similar to the real text features seems to be extremely low, because the similarity within the modality is greater than the similarity between the image and text pairs.
> 
> I also observed that the perturbation added at this time seems to be very large. Is the reason for the large perturbation here? If the text features and image features are highly aligned in the semantic space, is there no need for such a large perturbation?


The probability is low, that's why there is a performance gap between supervised method and language-free method. I will upload a preprint related to this problem to arxiv later. 

As I mentioned earlier, the hyper-parameter for perturbation depends on the pre-trained CLIP model. You can check a previous discussion here https://github.com/drboog/Lafite/issues/9#issuecomment-1086793768. If we have a pre-trained CLIP-like model that outputs image-text similarity close to 1 for ground-truth pairs, then I believe we can get better results with language-free method.

**Comment by drboog** [comment]
> Really thank you for your patient explanation! Now there's a lot of clarity on this! So maybe by using a "Gap-free" CLIP would increase the language-free training efficiency? Since we dont need to add such big noise. Anyway, hope you success in your future work!

Yes, a better pre-trained model will definitely improve the performance.

**Comment by Cwj1212** [comment]
Thank you so much for your answer, I don't even know how to thank you. My classmate  @EmiliaKKK  and I have been struggling with this problem. Thank you very much for the paper you recommended to us, this paper is very useful to us and we hadn't found it before. You are so kind that you can always answer our beginner's questions promptly and help us learn about your work. Thanks again!

**Comment by Cwj1212** [comment]
You already explained it in https://github.com/drboog/Lafite/issues/9#issuecomment-1086793768, I'm sorry to ask you again here without looking at someone else's issue before asking you

**Comment by Cwj1212** [comment]
I have a new conjecture about CLIP's multimodal semantic space.

Although the matching text-image pair has only 0.3 similarity, the difference between them in high-dimensional space is more similar to noise, so huge noise will not change the semantics. The matched text features and image features fluctuate around a 512-dimensional vector, and they all represent the same semantics. (For example, the vectors (2, 2, 2, 2) and (4, 4, 4, 4) are not very similar, but they all fluctuate around the center of the vector (3, 3, 3, 3). The fluctuating ones represent similar semantics)

While the similarity within the modality is as high as 0.5, their similarity is not as uniform as the noise. (For example, the vectors (2,2,2,2) and (2,2,2,6) although the distance between them is not particularly large, but the semantics are indeed quite different. (2,2,2,2) and The semantics of (3,3,3,3) are closer, and the semantics of (2,2,2,6) are closer to (1,1,1,5).

When you perturb the image with noise, although the vector changes a lot, the semantics don't change much. Inside the modal, although the vector changes little, the semantics changes a lot.

I don't know if you agree with this explanation @drboog

**Comment by Cwj1212** [comment]
Is it possible to make such a guess that after the image features are perturbed, the similarity between them and the image features is about 0.3, and the distribution of random pseudo-text labels can be regarded as a sphere around the image feature points, while the real text Labels may only be part of this sphere. Even if the similarity between image features and other image features is as high as 0.5, since this is a high-dimensional space, the spheres corresponding to these two image features do not overlap very much. We train the features on this sphere to generate images that are semantically consistent with the image features, and the text features are on top of this sphere.

**Comment by Cwj1212** [comment]
If the above assumption is correct. Because I can't understand the high-dimensional space well, is it because the dimension is too high, the spherical space after the perturbation of the image features is actually difficult to include the real text space (but this can rely on a large amount of data to continuously train to be solved), and the spherical space formed by different image features will not easily coincide (which makes the training successful).

**Comment by Cwj1212** [comment]
Maybe this is enough for a beginner like me, you mentioned above that you will upload a preprint related to this problem to arxiv later.If you will explain my above question in it, I will not go ahead and ask you for spoilers here haha. Looking forward to your follow-up work. thank you very much!

**Comment by drboog** [comment]
> Maybe this is enough for a beginner like me, you mentioned above that you will upload a preprint related to this problem to arxiv later.If you will explain my above question in it, I will not go ahead and ask you for spoilers here haha. Looking forward to your follow-up work. thank you very much!

You're welcome. In that paper I will talk about what better language-free methods we can design based on this problem. We can talk privately if you are interested in the quantitative results (not method itself), but I will not post the results here.
